### ISSUE TITLE
How do I cancel the work of inner producers?

### ISSUE THREAD

**Issue by bdolman** [ReactiveCocoa/ReactiveCocoa#2453]
Let's say I have parallel work I want to do, such as downloading multiple images. Each of these downloads is represented by a SignalProducer. I create them and flatten them using the Merge strategy. When I start the outer producer and get a disposable I expected that disposing of that would also dispose of the inner producers. What actually happens is that after disposal, my outer producer stops receiving events, but the actual work being done by the inner producers continues, because they have not been disposed.

How do I get the actual work being done by the inner producers to stop? Here's some test code that illustrates the scenario:

```swift
func producerOfValues(values: [String]) -> SignalProducer<String, NoError> {
    return SignalProducer<String, NoError> {observer, disposable in
        for value in values {
            sleep(1)
            print("doing work: \(value)")
            sendNext(observer, value)
            if disposable.disposed { // 'disposed' is never true
                print("producer disposed")
                break
            }
        }
        sendCompleted(observer)
    }
}

let (signal, sink) = SignalProducer<SignalProducer<String, NoError>, NoError>.buffer(2)
let flattened = signal.flatten(.Merge)

let letterProducer = producerOfValues(["a", "b", "c"])
let numberProducer = producerOfValues(["1", "2", "3"])

sendNext(sink, letterProducer.startOn(QueueScheduler()))
sendNext(sink, numberProducer.startOn(QueueScheduler()))
sendCompleted(sink)

let disposable = flattened.startWithNext {
    print($0)
}
disposable.dispose()
```

Output:

```text
doing work: a
doing work: 1
doing work: b
doing work: 2
doing work: c
doing work: 3
```

**Comment by NachoSoto** [comment]
Hmm this seems like a bug. I would expect the disposal to cancel the inner producers as well, but I'm not 100% sure. @andy_matuschak do you know if this is the expected behavior? It feels wrong.

**Comment by andymatuschak** [comment]
Agreed: that's how `-[RACSignal flatten:]` worked.

**Comment by stevebrambilla** [comment]
I was about to create a new issue before seeing this one, but I think what I've been digging into this afternoon is the same underlying issue.

After updating my RAC dependency I noticed that some of my tests were failing due to inner producers continuing to be started after their outer producers have been disposed of (due to a `take(1)` or similar).

I've been using this test case to debug the issue:

```swift
it("should not start further inner producers after outer producer is disposed") {
    let (firstInnerProducer, firstInnerSink) = SignalProducer<Int, NoError>.buffer()

    var secondInnerStarted = false
    let secondInnerProducer = SignalProducer<Int, NoError> { observer, _ in
        secondInnerStarted = true
        sendCompleted(observer)
    }

    let outerProducer = SignalProducer(values: [firstInnerProducer, secondInnerProducer])
        .flatten(.Concat)
        .take(1)

    var completed = false
    outerProducer.startWithCompleted {
        completed = true
    }

    sendNext(firstInnerSink, 1)
    sendCompleted(firstInnerSink)

    expect(completed).to(beTrue())
    expect(secondInnerStarted).to(beFalse())
}
```

The issue was introduced when `SignalProducer`'s flatten operators were changed to use `lift`. This test passes on 6b617c0, which is the last commit before that change, but fails on anything more recent.

The Interrupted event is not making it from the lifted signal to the transformed flatten signal. I haven't quite narrowed in on exactly why just yet. I can keep digging into this next week, but I'm hoping that someone with a better understanding of the intricacies of signal disposal can help shed some light on the issue :)

**Comment by bdolman** [comment]
I can confirm the regression describe by @stevebrambilla. My code behaves as expected when using 6b617c0. I tried to debug it and figure out what was going on but got pretty lost in the RAC innards.

**Comment by bdolman** [comment]
It seems the specific commit in which this broke is 100069f

**Comment by NachoSoto** [comment]
Aha! That would make sense. I'll try to work on a test case to reproduce this so that it doesn't break again.

**Comment by bdolman** [comment]
:metal: Thanks!

**Comment by stevebrambilla** [comment]
I've taken deeper look into this and the issue is a bit complex. I think what's happening here is due to:

1. The way `flatten` is expected to work is that when its base `Signal` completes, it finishes its outstanding work from the inner-signals before forwarding the `Completed` event.

2. The way that `Signal` _termination_ works is that as soon as a terminating event is sent to its observers, it removes the set of observers so no further events are propagated and the signal can eventually be deallocated.

3. The way that `Signal` _disposal_ works is by sending the `Interrupted` event to its observers, which propagates the disposal downstream.

This problem occurs when the base signal (the signal-of-signals) completes naturally, and _then_ you try to cancel everything (including the inner signals) by disposing it. The base signal's `Completed` event doesn't terminate the signal because of (1). Then because of (2), the `Interrupted` event from (3) isn't propagated downstream.

This test case is a more isolated demonstration of this than my last one. When `producer.start()` is called, the `baseProducer`'s signal immediately completes, rendering it impossible for the `Interrupted` event from `dispose()` to make it to the transformed signal.

```swift
it("should interrupt transformed signal when disposing of lifted signal") {
    // Empty base producer completes immediately
    let baseProducer = SignalProducer<Int, NoError>.empty

    // Expect an Interrupted event in order to propagate disposal
    var wasInterrupted = false

    let transform = { (signal: Signal<Int, NoError>) -> Signal<Int, NoError> in
        return Signal { observer in
            return signal.observe { event in
                switch event {
                case .Completed:
                    // "Defer" Completed events to simulate `flatten()` waiting for work to finish
                    break

                case .Interrupted:
                    // We would need the Interrupted event to dispose of our work if it was `flatten()`
                    wasInterrupted = true
                    observer.sendInterrupted()

                default:
                    observer.action(event)
                }
            }
        }
    }

    let producer = baseProducer.lift(transform)
    let disposable = producer.start()

    disposable.dispose()
    expect(wasInterrupted).to(beTruthy())
}
```

When `flatten()` was implemented as a `SignalProducer` operator this wasn't an issue because the producer's disposable would be disposed of directly.

How do you think we should proceed here? I'm not yet sure how to make `flatten()` disposal work this way as a `Signal` operator without doing some pretty hacky things..

**Comment by schickling** [comment]
Looking forward to a solution of this!

**Comment by neilpa** [comment]
This makes me more inclined to look into #2044. However the fallout is going to be large.

**Comment by mhuusko5** [comment]
I'm having a few similar experiences where inner disposables are unreliably (or not at all) being disposed of when outer ones are disposed of, *and vice versa*, when using `.flatMap(.Latest)`.

This is blocking me atm, and I was wondering if anyone has a short term solution/a drop in to get the pre-https://github.com/ReactiveCocoa/ReactiveCocoa/commit/100069f7abc0bafaaf46525d75035c9f032bf999 behaviour?

**Comment by NachoSoto** [comment]
A short-term solution could be to insert explicit `takeUntil` calls before and/or after `flaten`ing.

I'll try to get to this this weekend!

**Comment by mhuusko5** [comment]
Oh, interesting. Thanks!

**Comment by mhuusko5** [comment]
Btw, wanting to check if this was related to the same problem:

```swift
firebase.cancelDisconnectOperations()
    .then(firebase.userSession(userID: joinUserID).observeEvent(.Value))
    .takeWhile { $0.isLive }
    .on(disposed: {
        log("This gets disposed of when takeWhile returns false! Yay!")
    })
    .flatMap(.Latest) { joinedSession -> SignalProducer<Session, Error> in
        let headUserID = joinedSession.joinedItem.headUserID
        
        return firebase.userSession(userID: headUserID).observeEvent(.Value)
    }
    .on(disposed: {
        log("This only gets called if flatMap was never reached (takeWhile returned false first time). Otherwise this never gets called.")
    })
```

Once takeWhile returns false, I would expect a 'completed' event to travel down the top level producer, instead, conceptually it seems to be trapped by the flatMap.

**Comment by NachoSoto** [comment]
So you only see `"This gets disposed of when takeWhile returns false! Yay!"`? That's expected.
`takeWhile` is stopping the outer `producer`, but an inner `producer` that's currently running will continue. You'd need `takeWhile` or `takeUntil`, or something similar **after** `flatMap`.

Does that make sense?

**Comment by mhuusko5** [comment]
It doesn't make sense, but it's a learning process ;)

So, an interruption of an outer producer isn't even *supposed* to lead to an interruption down the chain/of mapped producers?

**Comment by neilpa** [comment]
[The process of `flatten`ing creates a new producer that requires both the outer an inner producers to terminate in a non-failure event](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/49d75631c0daf4c470cb5d9c5efb8b1d19aab05c/ReactiveCocoa/Swift/Signal.swift#L293-L294). That's why the `take*` operator should happen after the `flatMap` above.

**Comment by mhuusko5** [comment]
Ah, okay, that makes sense.
I'm wondering how to achieve the above then (without adding a bunch of hacky state)? I need to short circuit the chain if what the outer producer returns eventually turns isLive=false. Adding the takeUntil after the flatMap would be acting on an irrelevant resource down the chain.

**Comment by mhuusko5** [comment]
I've got very close to I want, but have encountered another oddity with flatMap that would be good to know whether bug or expected behaviour:

```swift
firebase.cancelDisconnectOperations()
    .then(SignalProducer { sink, disposable in
        firebase.userSession(userID: joinUserID).observeEvent(.Value, model: Session.self)
            .takeWhile { $0.isLive }
            .on(completed: {
                sink.sendInterrupted() // 1) This gets called
            })
            .flatMapLatest { joinedSession in
                let headUserID = joinedSession.joinedItem?.headUserID ?? joinUserID
                
                return userSession.updateChildValues(["head": headUserID])
                    .then(firebase.userSession(userID: headUserID).observeEvent(.Value))
            }
            .start(sink)
            .addToDisposable(disposable)
    })
    .on(disposed: {
        // 2) This gets called, as expected
    }, interrupted: {
        // 3) This should get called, but doesn't
    }, completed: {
       // 4) This gets called
    })
```

In this case I specifically want to interrupt the signal, because I respond to an interrupt differently than success/complete downstream, but sending an interrupt, although does all the cleanup I was looking for, actually completes the signal.

**Comment by mdiep** [comment]
Okay, I've spent a bunch of time looking at this. I have some understanding of what the issue is, but not how to fix it… which is why I'm commenting here instead of opening a PR.

I've been working with this altered version of the original code snippet:

```swift
					let (signal, sink) = SignalProducer<SignalProducer<Int, NoError>, NoError>.buffer(2)
					let flattened = signal.flatten(.Concat)
					
					var completed = false
					let values = [1, 2, 3, 4, 5]
					let numberProducer = SignalProducer<Int, NoError> {observer, disposable in
						sleep(10)
						for value in values {
							if disposable.disposed { // 'disposed' is never true
								print("producer disposed")
								break
							}
							print("doing work: \(value)")
							observer.sendNext(value)
						}
						observer.sendCompleted()
						completed = true
					}
					
					let scheduler: QueueScheduler
					if #available(OSX 10.10, *) {
						scheduler = QueueScheduler(qos: QOS_CLASS_DEFAULT)
					} else {
						scheduler = QueueScheduler(queue: dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0))
					}
					sink.sendNext(numberProducer.startOn(scheduler))
					sink.sendCompleted()
					
					let disposable = flattened.start()
					disposable.dispose()
					
					expect(completed).toEventually(beTruthy(), timeout: 100)
```

My understanding is that this _should_ print this:

```
producer disposed
```

But instead, you get this:

```
doing work: 1
doing work: 2
doing work: 3
doing work: 4
doing work: 5
```

However, if you remove the `sink.sendCompleted()`, then it _is_ disposed as expected.

This is because the inner signals are only disposed if the sink propagates an Interrupted event down the chain of observers. The `concat` will see the Interrupt and interrupt its observer, which leads to the disposal of the inner signals. But because the sink is completed, the interrupt doesn't propagate.

To fix it, the disposal of the flatten producer needs to lead to the disposal of the concatted signal, which I think needs to send the interrupt to the concat's observer. But I'm struggling to understand how it's supposed to work, so it'd be helpful if someone else could jump in. Hopefully my notes will simplify things.

**Comment by cradnovich** [comment]
After looking at this for some time, I have gotten to the same place as @mdiep. I wonder whether this is really a bug or the expected behaviour of `Signal`s as something that is described in the documentation as "represent[ing] event streams that are already 'in progress', like notifications, user input." It seems consistent that completing the outer signal-of-signals works like it does. Making the Signal handle SignalProducer or Signal values differently from any other Value type might work for this but it doesn't seem right. I'll check out the old version for a similar test to see what the difference is.

**Comment by neilpa** [comment]
After more thought I've realized that it's impossible to implement `flatten` for producers-of-producers in terms of `lift`. As @cradnovich is getting at, the semantics of disposing a `Signal` observation is unsubscribe, not cancel internal work. We're going to need explicit implementations of any `flatten` operator for both producers-of-producers and signals-of-producers.

**Comment by neilpa** [comment]
Also, `flatMapError` has this same problem.

**Comment by NachoSoto** [comment]
Maybe we can find a way to implement an overload of lift that disposes the started producer upon unsubscription?

Otherwise, having a separate implementation would be unfortunate, but I guess it would make sense.

**Comment by neilpa** [comment]
> Maybe we can find a way to implement an overload of lift that disposes the started producer upon unsubscription?

That won't work if there are multiple subscribers to the signal. Basically once a signal-of-producers starts embedded work it can no longer be canceled. The ability to cancel inner work only makes sense for an producer-of-producers

**Comment by NachoSoto** [comment]
Right, we'd have to keep a count of subscribers and only dispose when it gets to 0.

**Comment by neilpa** [comment]
That's still not right because someone could still subscribe to a `Signal` after it goes to 0.

**Comment by NachoSoto** [comment]
I know, that's why I'm only talking of doing this for lifting flatten.

**Comment by neilpa** [comment]
But how would that work? I don't see a way to expose this information out of a `Signal` operator so it can be consumed by `lift`.

**Comment by NachoSoto** [comment]
We'd have to wrap the `Signal` so that we can keep track of subscribers, which might not be pretty. I don't know, I haven't fully thought about it, and it's possible that maintaining separate implementations is still simpler. I'm just throwing out ideas because I'd like to avoid that if we can.

**Comment by NachoSoto** [comment]
We could make the canonical implementation take a closure for when an inner signal is unsubscribed, so the `SignalProducer` version can dispose.

**Comment by NachoSoto** [comment]
Of course hiding that from the public API, just as part of the canonical shared implementation of the flatten operators.

**Comment by Vadim-Yelagin** [comment]
Let me try to elaborate this.

The way I see it, a signal is something that is already happening, and you can only observe (subscribe) and unsubscribe without affecting the underlying process. A signal producer is something that will start happening when you `start` it, and it will also produce a signal (hence the name) corresponding to that process.

According to this logic the result of flattening a signal of signal producers cannot be a signal, because observing this flattened "signal" will result in starting the inner signal producers. One would also expect that disposing (unsubscribing) from it will not only interrupt the already ongoing inner producers (which does not happen per this issue) but also prevent the future producers from being started (as a result of this observation of the flattened "signal"). Both this points contradict the requirement that observing a signal has no effect on the underlying process, and that "disposing a Signal observation is unsubscribe".

What it means is that the result of flattening a signal of signal producers can only be a signal producer. In terms of implementation it means that there have to be two different implementations for a signal of signals (returning a signal) and for a producer of producers (returning a producer).
The implementation for signal of producers (returning a producer) can be made from the implementation for producer of producers by wrapping the outer signal in a producer.
The implementation for producer of signal (returning a producer) can be made by lifting the implementation for signal of signals.

**Comment by Vadim-Yelagin** [comment]
Moreover, flattening of signal of signal producers cannot be a signal because "an observer of a Signal will see the exact same sequence of events as all other observers", but for each observer the inner signal producers will be started anew and yield different sequences of events.

**Comment by NachoSoto** [comment]
I think you're absolutely right: `Signal<SignalProducer<T, E>, E>.flatten` should return a `SignalProducer<T ,E>` /cc @neilpa.

**Comment by neilpa** [comment]
I think that you guys are forgetting that every signal operator immediately subscribes the source signal when it's applied. In the case of `flatten` that means we set up a single subscription on the underlying `Signal<SignalProducer<T, E>, E>`. Each time the source signal sends a new producer it's started immediately and exactly once. This is independent from any subscriptions on the `flatten`ed signal.

**Comment by neilpa** [comment]
Here's an example that hopefully illustrates my point and shows why signal-of-producers can return a signal.

```swift
let (signal, signalPipe) = Signal<SignalProducer<String, NoError>, NoError>.pipe()
let flattened = signal.flatten(.Merge)

var producerObserver: Observer<String, NoError>!
let producer = SignalProducer<String, NoError> { observer, disposable in
    producerObserver = observer
    print("Starting producer") // only started once

    disposable.addDisposable {
        print("Should be called last")
    }
}

signalPipe.sendNext(producer)
producerObserver.sendNext("First") // no one sees this

let d = flattened.observeNext { print("observer 1: \($0)") }
producerObserver.sendNext("Second")

flattened.observeNext { print("observer 2: \($0)") }
producerObserver.sendNext("Third")

signalPipe.sendNext(SignalProducer(values: ["A", "B"]))

d?.dispose()
producerObserver.sendNext("Fourth")

print("Signal completing")
signalPipe.sendCompleted()

print("Producer completing")
producerObserver.sendCompleted()
```

And produces the following output

```
Starting producer
observer 1: Second
observer 1: Third
observer 2: Third
observer 1: A
observer 2: A
observer 1: B
observer 2: B
observer 2: Fourth
Signal completing
Producer completing
Should be called last
```

**Comment by Vadim-Yelagin** [comment]
@neilpa You mean, the producers get started as long as the flattened signal exists as an object?

**Comment by neilpa** [comment]
@Vadim-Yelagin Correct, that's the case for any signal. As long as it's alive it's sending events.

**Comment by Vadim-Yelagin** [comment]
@neilpa how does it stop the new producers from being started (and interrupts the producers that are already running) after it deallocs?

**Comment by neilpa** [comment]
What are you're asking isn't possible. If it deallocs how is it going to see new producers to start?

**Comment by Vadim-Yelagin** [comment]
@neilpa I mean, once you create this flattened signal, it starts to observe the original signal and it starts every producer it emits, and it stops only after the original signal finishes. You cannot stop this sooner even if you never observe this flattened signal and throw it away immediately after you create it. Right?

**Comment by neilpa** [comment]
Yes, that's fundamentally how `Signal` works. It's an always on stream of values [whose lifetime is tied to those events](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/4faa7bc75cd30682e9e5615bb6359988f0339800/ReactiveCocoa/Swift/Signal.swift#L15-L17).

**Comment by neilpa** [comment]
> Of course hiding that from the public API, just as part of the canonical shared implementation of the flatten operators.

I experimented with this some last night but haven't figured out a way to do this universally. It is possible to share portions of the individual implementations across the different `flatten` operators though.

**Comment by NachoSoto** [comment]
> It is possible to share portions of the individual implementations across the different flatten operators though.

That would be great!

**Comment by NachoSoto** [comment]
Fixed!

**Comment by bdolman** [comment]
Thanks @neilpa @NachoSoto and everyone else who helped on this. It's working great now.

**Comment by NachoSoto** [comment]
😍
